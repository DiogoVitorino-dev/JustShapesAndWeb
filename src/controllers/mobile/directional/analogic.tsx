import React from "react";
import { View, StyleSheet } from "react-native";
import { Gesture, GestureDetector } from "react-native-gesture-handler";
import Animated, {
  Easing,
  useAnimatedStyle,
  useSharedValue,
  interpolate,
  useAnimatedReaction,
  withTiming,
} from "react-native-reanimated";

import { JoystickData, JoystickProp } from "..";

import { AnglesUtils } from "@/scripts/utils/angleUtils";

interface AnalogicProp extends JoystickProp {}

export default function AnalogicJoystick({
  size = 100,
  velocity = 10,
  onMove,
  containerStyle,
}: AnalogicProp) {
  const radius = size / 2;
  const pointerSize = size / 2.5;
  const pos = useSharedValue({ x: 0, y: 0 });

  const normalizeData = ({ angle, x, y }: JoystickData): JoystickData => {
    "worklet";

    // adjusting to velocity range
    x = interpolate(x, [radius * -1, radius], [velocity * -1, velocity]);
    y = interpolate(y, [radius * -1, radius], [velocity * -1, velocity]);

    return { angle, x, y };
  };

  const updatePointerPosition = (x: number, y: number) => {
    "worklet";
    pos.value = { x, y };
  };

  const pan = Gesture.Pan()
    .onBegin(({ x, y }) => {
      updatePointerPosition(x - radius, y - radius);
    })

    .onUpdate(({ x, y }) => {
      const posX = x - radius;
      const posY = y - radius;

      // Distance from center to the edge of the area
      const distance = Math.sqrt(Math.pow(posX, 2) + Math.pow(posY, 2));

      // Limits the pointer to the edge of the area
      if (distance >= radius) {
        const factor = radius / distance;
        updatePointerPosition(posX * factor, posY * factor);
      } else {
        updatePointerPosition(posX, posY);
      }
    })

    .onFinalize(() => {
      updatePointerPosition(0, 0);
    })
    .withTestId("joystickAnalogicPan");

  const pointerAnimatedStyle = useAnimatedStyle(() => ({
    transform: [
      {
        translateX: withTiming(pos.value.x, {
          duration: 20,
          easing: Easing.out(Easing.ease),
        }),
      },
      {
        translateY: withTiming(pos.value.y, {
          duration: 20,
          easing: Easing.out(Easing.ease),
        }),
      },
    ],
  }));

  useAnimatedReaction(
    (): JoystickData => {
      const angle = AnglesUtils.calculateAngle(pos.value.x, pos.value.y);
      return normalizeData({
        angle,
        x: pos.value.x,
        y: pos.value.y,
      });
    },
    (currentValue, previousValue) => {
      if (currentValue !== previousValue && onMove) {
        onMove(currentValue);
      }
    },
  );

  return (
    <View style={[styles.container, containerStyle]}>
      <GestureDetector gesture={pan}>
        <View
          testID="pan"
          style={[
            { width: size, height: size, borderRadius: radius },
            styles.area,
          ]}
        >
          <Animated.View
            style={[
              {
                width: pointerSize,
                height: pointerSize,
                borderRadius: radius,
              },
              styles.pointer,
              pointerAnimatedStyle,
            ]}
          />
        </View>
      </GestureDetector>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    width: "30%",
    height: "50%",
    alignItems: "center",
    justifyContent: "center",
    backgroundColor: "indigo",
  },

  area: {
    backgroundColor: "tomato",
    alignItems: "center",
    justifyContent: "center",
    zIndex: 1000,
  },

  pointer: {
    backgroundColor: "cyan",
  },
});
