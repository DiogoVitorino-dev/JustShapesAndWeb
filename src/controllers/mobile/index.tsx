import React from "react";
import { StyleSheet, View, ViewStyle } from "react-native";
import { useAnimatedReaction, useSharedValue } from "react-native-reanimated";

import AreaButton from "./buttons/area";
import AnalogicJoystick from "./directional/analogic";
import { ControlData, ControlProp, Directional } from "../typesControllers";

interface MobileControlProp extends ControlProp {}

export interface JoystickData extends Directional {}

export interface JoystickProp {
  size?: number;
  velocity?: number;
  onMove?: (data: JoystickData) => void;
  containerStyle?: ViewStyle;
}

export default function MobileControl({ onMove, velocity }: MobileControlProp) {
  const jump = useSharedValue<boolean>(false);
  const data = useSharedValue<JoystickData>({ angle: 0, x: 0, y: 0 });

  const handleOnMove = (newData: JoystickData) => {
    "worklet";
    data.value = newData;
  };

  const handleOnJump = (pressed: boolean) => {
    "worklet";
    jump.value = pressed;
  };

  useAnimatedReaction(
    (): ControlData => ({ jumping: jump.value, ...data.value }),
    (current, previous) => {
      if (current !== previous) {
        onMove(current);
      }
    },
  );

  return (
    <View style={styles.container}>
      <AnalogicJoystick onMove={handleOnMove} velocity={velocity} />
      <AreaButton onPress={handleOnJump} />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    width: "100%",
    flex: 1,
    flexDirection: "row",
    alignItems: "flex-end",
    justifyContent: "space-between",
  },
});
