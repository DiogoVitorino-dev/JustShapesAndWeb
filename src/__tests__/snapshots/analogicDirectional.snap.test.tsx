import { render } from "@testing-library/react-native";

import { AnalogicDirectional } from "@/controllers/mobile/directional";

describe("Analogic Directional - Snapshot mobile controller test", () => {
  it("should render correctly", () => {
    const tree = render(
      <AnalogicDirectional onMove={() => {}} size={300} velocity={50} />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
