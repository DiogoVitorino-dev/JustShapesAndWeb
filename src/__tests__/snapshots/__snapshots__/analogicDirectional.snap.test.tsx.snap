// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Analogic Directional - Snapshot mobile controller test should render correctly 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "rgb(3, 38, 25)",
      },
      [
        {
          "alignItems": "center",
          "height": "50%",
          "justifyContent": "center",
          "width": "30%",
        },
        undefined,
      ],
      {
        "backgroundColor": "transparent",
      },
    ]
  }
>
  <View
    collapsable={false}
    style={
      [
        {
          "backgroundColor": "rgb(3, 38, 25)",
        },
        [
          {
            "borderRadius": 150,
            "height": 300,
            "width": 300,
          },
          {
            "alignItems": "center",
            "backgroundColor": "rgb(8, 78, 94)",
            "elevation": 3,
            "justifyContent": "center",
            "zIndex": 1000,
          },
        ],
      ]
    }
    testID="pan"
  >
    <View
      animatedStyle={
        {
          "value": {
            "transform": [
              {
                "translateX": 0,
              },
              {
                "translateY": 0,
              },
            ],
          },
        }
      }
      collapsable={false}
      style={
        {
          "backgroundColor": "rgb(20, 210, 255)",
          "borderRadius": 150,
          "height": 120,
          "transform": [
            {
              "translateX": 0,
            },
            {
              "translateY": 0,
            },
          ],
          "width": 120,
        }
      }
    />
  </View>
</View>
`;
